### ISSUE TITLE
onMessage sometimes not called when typing enabled

### ISSUE THREAD

**Issue by hitchingsh** [pusher/chatkit-swift#114]
### What?

We are developing an app using Chatkit to do chat on iOS.  onMessage was being called reliably until we enabled the typing indicator feature for Chatkit in swift a couple of months ago.  We found that every certain number of messages (say about 1 or 2 in 10) the onMessage would not be called.  This has happened to both of on an ongoing basis.  We are located in different countries so it is not a network issue.  Calling self.currentUser.typing function is what prevents on message from sometimes being called:

    func messageInputBar(_ inputBar: MessageInputBar, textViewTextDidChangeTo text: String) {
        
        self.currentUser.typing(in: self.room) { error in
            NSLog("messageInputBar::textViewTextDidChangeTo text=\(text)")
            if ((error) != nil) {
                NSLog("Error: messageInputBar: \(String(describing: error))")
            }
        }
    }

    public func onMessage(_ message: PCMessage) {
        NSLog("onMessage Received message with msgId: \(message.id) for roomId: \(room.id) text: \(message.text) from \(message.sender.debugDescription)")
        
        self.addMessageToUI(message)
    }

Also, can you let me know how I can set the debugging messages to verbose?

### Environment 
MacBook pro

* **PusherChatkit version:** 0.6.2
* **Xcode version:** 10.1
* **Swift version:** 11
* **Platform(s) running PusherChatkit:** 1.1.0

### Demo Project (if you have one; it helps a lot!)

----

CC @hamchapman

**Comment by hamchapman** [comment]
@hitchingsh I've tried to recreate this using this super simple app I made: https://github.com/hamchapman/ChatMessageKit

So far I've not been able to, following the actions you said worked for you.

It assumes that the users you connect with are only in one shared room, if you want to test it yourself. 

If you're able to share your app with me where you are able to recreate it then please share that with me so that we can debug further.

**Comment by hitchingsh** [comment]
It appears when we send a message with sendMessage, the message is not being sent back from the Chatkit/Pusher server sometimes when the user is typing.  Specifically, PusherPlatform::PPSubscriptionURLSessionDelegate class with method urlSession is not being called in the cases where onMessage is not called. In other words, I believe this is a Pusher server issue. I inserted NSLog statements in the PusherPlatform code to determine this. I also noticed sometimes when the app is in the background and then brought to the foreground sometimes onMessage is not called.

If I type and send messages in our app within the first ten and usually the first five messages I send it skips calling onMessage. Our code for typing appears is identical to yours as shown below.

    func messageInputBar(_ inputBar: MessageInputBar, textViewTextDidChangeTo text: String) {

       self.currentUser.typing(in: self.room) { error in
            NSLog("messageInputBar::textViewTextDidChangeTo text=\(text)")
            if ((error) != nil) {
                NSLog("Error: messageInputBar: \(String(describing: error))")
            }
            print("Is typing")
        }
    }

I downloaded and tested your app and your app works where onMessage is always called even with the typing set identical to ours as shown above and our app stops calling onMessage when we turn on the typing indicator. I've also ensured all libraries and OS versions are the same. However, I don't have the access to debug why your server is not returning a message to the client after we call sendMessage.

**Comment by hamchapman** [comment]
Given that the code in the demo app I shared and the code you're running seems to be similar, it sounds like the issue is localised to your app, although I'm hesitant to rule anything out just yet!

When you don't see `onMessage` being called, are you completely sure that the message has successfully sent? Could it be that the message send is failing, rather than the message delivery? If you reload the app and navigate to the same room, is the message then delivered?

**Comment by hitchingsh** [comment]
The message is always successfully sent and stored on the chatkit server 100% of the time.  When I go back into the room the messages are there.  Please look at the server logs for the instance locators I gave you and see if there are any errors and also if chatkit does not send the message back to you if you are typing or if it drops messages

**Comment by hamchapman** [comment]
I've just checked our server logs and checked our monitoring to see if there were any errors related to your instance locator and there haven't been.

Are you able to share your app with me where you're experiencing the issue, either in its full form or in a stripped down form?

Otherwise I'm not sure what else I can do to help debug if neither you nor I can reproduce the bug in the app I made and linked to above, which seems to be fairly similar to what you've got.

Thanks!

**Comment by hitchingsh** [comment]
Ham Chapman, I've sent you a Github invite to our repro so you can download it and test it to reproduce the problem of onMessage not being called when sending messages with the typing indicator turned on.   Please do not share this code. 

The default points to our test server. Login using the default number 1-650-555-5555 and enter the six digit verification code (shown in XCode console for development purposes).  Then start typing messages like "Hi1" in the chat.  Before you get to "Hi10" you will notice onMessage is not called.

For your debugging purposes, all the action is in RoomVC.swift including the onMessage call and  self.currentUser.typing(in: self.room) { error in 

The server always receives and stores the message but sometimes when the user is typing the message is not sent back to the client so there is a good chance its a server side problem.

**Comment by hitchingsh** [comment]
Note you have to use the SLKInputBar branch in github

**Comment by hamchapman** [comment]
@hitchingsh I'm not 100% sure what I need to do to recreate the bug. I've tried typing messages in one simulator while then sending a message from another one and everything seems to be fine. 

Are you able to provide a video or some written steps that explains how to trigger the problem?

So far every message I've sent seems to have been received by all connected clients every time so I suspect I'm not doing the right thing to recreate the problem.

**Comment by hitchingsh** [comment]
Make sure you pulled SLKInputBar branch.  Also, in RoomVC make sure its uncommented. Then type in 10 messages in a row like "Hi1", "Hi2"

    func messageInputBar(_ inputBar: MessageInputBar, textViewTextDidChangeTo text: String) {
        
        // TODO: Typing indicator
        self.currentUser.typing(in: self.room) { error in
            // NSLog("messageInputBar::textViewTextDidChangeTo text=\(text)")
            if ((error) != nil) {
                NSLog("Error: messageInputBar: \(String(describing: error))")
            }
            print("Is typing")
        }
    }

**Comment by hitchingsh** [comment]
Thank you. That works for our development schedule.

**Comment by hamchapman** [comment]
The work for the fix is ongoing and approaching completion. Apologies that it's taken a while to get it all sorted.

In terms of an ETA, I'd estimate that it should be live by the end of next week, give or a take a bit.

**Comment by clay-mindoula** [comment]
That works, thankee

**Comment by ZakSingh** [comment]
Ran into this in our react native app - drove us crazy. Glad there is a fix in progress!

**Comment by vivangkumar** [comment]
Hi all!

Apologies for the lack of communication on the thread, we have been hard at work fixing this. The issue has taken a significant amount of time, substantially more than what we anticipated. It has taken a few iterations to get the solution right, and the fix is currently being reviewed + rigorously tested. We'll deploy it as we've assured ourselves we've fixed it for good. Hopefully end of this week, if not, early next week. Thank you all for your patience.
